pkg/ring/lifecycler.go

```diff
@@ -504,23 +517,32 @@ func (i *Lifecycler) initRing(ctx context.Context) error {
 
 		ingesterDesc, ok := ringDesc.Ingesters[i.ID]
 		if !ok {
+			// The instance doesn't exist in the ring, so it's safe to set the registered timestamp
+			// as of now.
+			registeredAt := time.Now()
+			i.setRegisteredAt(registeredAt)
+
 			// We use the tokens from the file only if it does not exist in the ring yet.
 			if len(tokensFromFile) > 0 {
 				level.Info(util.Logger).Log("msg", "adding tokens from file", "num_tokens", len(tokensFromFile))
 				if len(tokensFromFile) >= i.cfg.NumTokens {
 					i.setState(ACTIVE)
 				}
-				ringDesc.AddIngester(i.ID, i.Addr, i.Zone, tokensFromFile, i.GetState())
+				ringDesc.AddIngester(i.ID, i.Addr, i.Zone, tokensFromFile, i.GetState(), registeredAt)
 				i.setTokens(tokensFromFile)
 				return ringDesc, true, nil
 			}
 
 			// Either we are a new ingester, or consul must have restarted
 			level.Info(util.Logger).Log("msg", "instance not found in ring, adding with no tokens", "ring", i.RingName)
-			ringDesc.AddIngester(i.ID, i.Addr, i.Zone, []uint32{}, i.GetState())
+			ringDesc.AddIngester(i.ID, i.Addr, i.Zone, []uint32{}, i.GetState(), registeredAt)
 			return ringDesc, true, nil
 		}
 
+		// The instance already exists in the ring, so we can't change the registered timestamp (even if it's zero)
```

Comment:
It may not be clear why we don't set "registered at" if it's zero. I would suggest extending  description of this field in ring.proto to describe its intended usage, and explaining why preserving zero is important.

Reply:
I updated the command. I haven't specifically referenced the shuffle sharding because it's not a thing yet and this is a generic property, but I tried to make it clear to never change it once its set (even if it's 0).

WDYT?

Reply:
What I meant is to explain purpose of the field. Something like: This field is used to find out subset of instances that could have possibly owned the specific token in the past.

Reply:
> What I meant is to explain purpose of the field. Something like: This field is used to find out subset of instances that could have possibly owned the specific token in the past.

And from there, explain why updating to non-zero value is a bad idea.

Reply:
Ok, updated!